File: app/api/schemas/rag.py
"""Authenticated-user contracts for personal knowledge resources."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class PageResponse(BaseModel):
    page: int
    page_size: int
    total: int


class KnowledgeBaseCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=10_000)

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        return value.strip()

    @field_validator("description")
    @classmethod
    def normalize_description(cls, value: str | None) -> str | None:
        normalized = value.strip() if value is not None else None
        return normalized or None


class KnowledgeBaseUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=10_000)
    status: Annotated[str | None, Field(pattern=r"^(active|disabled)$")] = None
    search_enabled: bool | None = None

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str | None) -> str | None:
        return value.strip() if value is not None else None

    @field_validator("description")
    @classmethod
    def normalize_description(cls, value: str | None) -> str | None:
        normalized = value.strip() if value is not None else None
        return normalized or None

    @model_validator(mode="after")
    def require_change(self) -> KnowledgeBaseUpdate:
        if not self.model_fields_set:
            raise ValueError("at least one knowledge-base field must be provided")
        return self


class KnowledgeBaseResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: str | None
    embedding_provider: str
    embedding_model: str
    embedding_dimensions: int
    embedding_distance_metric: str
    pipeline_version: str
    status: str
    search_enabled: bool
    created_at: datetime
    updated_at: datetime


class KnowledgeBaseListResponse(PageResponse):
    items: list[KnowledgeBaseResponse]


class RetrievalRequest(BaseModel):
    query: str = Field(min_length=1, max_length=2_000)
    limit: int = Field(default=10, ge=1, le=50)
    min_score: float | None = Field(default=None, ge=-1.0, le=1.0)

    @field_validator("query")
    @classmethod
    def normalize_query(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("query must contain non-whitespace characters")
        return normalized


class RetrievedChunkResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    chunk_id: UUID
    document_id: UUID
    document_version_id: UUID
    knowledge_base_id: UUID
    content: str
    title: str
    page_number: int | None
    section_path: str | None
    sheet_name: str | None
    row_start: int | None
    row_end: int | None
    char_start: int | None
    char_end: int | None
    metadata: dict[str, Any]
    score: float
    retrieval_source: str


class RetrievalResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    knowledge_base_id: UUID
    query: str
    embedding_model: str
    latency_ms: int
    items: list[RetrievedChunkResponse]


class DocumentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    knowledge_base_id: UUID
    name: str
    mime_type: str
    size_bytes: int
    content_hash: str
    status: str
    is_enabled: bool
    created_at: datetime
    updated_at: datetime


class DocumentListResponse(PageResponse):
    items: list[DocumentResponse]


class DocumentVersionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    document_id: UUID
    version: int
    content_hash: str
    status: str
    pipeline_version: str
    embedding_provider: str
    embedding_model: str
    embedding_dimensions: int
    error_code: str | None
    error_message: str | None
    metadata_json: dict[str, str]
    created_at: datetime


class DocumentVersionListResponse(BaseModel):
    items: list[DocumentVersionResponse]


class DocumentDownloadUrlResponse(BaseModel):
    url: str
    expires_at: datetime


class IngestionJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    document_id: UUID
    document_version_id: UUID
    status: str
    progress: int
    retry_count: int
    max_retries: int
    manual_retry_count: int
    error_code: str | None
    error_message: str | None
    error_detail: dict[str, Any] | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class IngestionJobListResponse(BaseModel):
    items: list[IngestionJobResponse]


class OutboxEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    ingestion_job_id: UUID
    status: str
    attempt_count: int
    max_attempts: int
    manual_retry_count: int
    available_at: datetime
    last_error: str | None
    published_at: datetime | None
    failed_at: datetime | None


class IngestionRetryResponse(BaseModel):
    ingestion_job: IngestionJobResponse
    dispatch_event: OutboxEventResponse


class DocumentUploadResponse(BaseModel):
    document: DocumentResponse
    version: DocumentVersionResponse
    ingestion_job: IngestionJobResponse

File: web/src/types/rag.ts
import type { PageResponse } from '@/types/api'

export type KnowledgeBaseStatus = 'active' | 'disabled'
export type DocumentStatus = 'uploaded' | 'published' | 'disabled' | 'deleted'
export type DocumentVersionStatus =
  | 'uploading'
  | 'awaiting_pipeline'
  | 'processing'
  | 'published'
  | 'superseded'
  | 'failed'
export type IngestionJobStatus =
  'awaiting_pipeline' | 'processing' | 'completed' | 'failed'
export type OutboxEventStatus = 'pending' | 'published' | 'failed'

export interface KnowledgeBase {
  id: string
  name: string
  description: string | null
  embedding_provider: string
  embedding_model: string
  embedding_dimensions: number
  embedding_distance_metric: string
  pipeline_version: string
  status: KnowledgeBaseStatus
  search_enabled: boolean
  created_at: string
  updated_at: string
}

export interface KnowledgeBaseList extends PageResponse {
  items: KnowledgeBase[]
}

export interface KnowledgeBaseInput {
  name: string
  description?: string | null
}

export interface KnowledgeBaseUpdate {
  name?: string
  description?: string | null
  status?: KnowledgeBaseStatus
  search_enabled?: boolean
}

export interface Document {
  id: string
  knowledge_base_id: string
  name: string
  mime_type: string
  size_bytes: number
  content_hash: string
  status: DocumentStatus
  is_enabled: boolean
  created_at: string
  updated_at: string
}

export interface DocumentList extends PageResponse {
  items: Document[]
}

export interface DocumentVersion {
  id: string
  document_id: string
  version: number
  content_hash: string
  status: DocumentVersionStatus
  pipeline_version: string
  embedding_provider: string
  embedding_model: string
  embedding_dimensions: number
  error_code: string | null
  error_message: string | null
  metadata_json: Record<string, string>
  created_at: string
}

export interface DocumentVersionList {
  items: DocumentVersion[]
}

export interface IngestionJob {
  id: string
  document_id: string
  document_version_id: string
  status: IngestionJobStatus
  progress: number
  retry_count: number
  max_retries: number
  manual_retry_count: number
  error_code: string | null
  error_message: string | null
  error_detail: Record<string, unknown> | null
  started_at: string | null
  completed_at: string | null
  created_at: string
  updated_at: string
}

export interface IngestionJobList {
  items: IngestionJob[]
}

export interface OutboxEvent {
  id: string
  ingestion_job_id: string
  status: OutboxEventStatus
  attempt_count: number
  max_attempts: number
  manual_retry_count: number
  available_at: string
  last_error: string | null
  published_at: string | null
  failed_at: string | null
}

export interface IngestionRetryResponse {
  ingestion_job: IngestionJob
  dispatch_event: OutboxEvent
}

export interface DocumentUploadResponse {
  document: Document
  version: DocumentVersion
  ingestion_job: IngestionJob
}

export interface DocumentDownloadUrl {
  url: string
  expires_at: string
}

export interface RetrievalInput {
  query: string
  limit?: number
  min_score?: number | null
}

export interface RetrievedChunk {
  chunk_id: string
  document_id: string
  document_version_id: string
  knowledge_base_id: string
  content: string
  title: string
  page_number: number | null
  section_path: string | null
  sheet_name: string | null
  row_start: number | null
  row_end: number | null
  char_start: number | null
  char_end: number | null
  metadata: Record<string, unknown>
  score: number
  retrieval_source: string
}

export interface RetrievalResponse {
  knowledge_base_id: string
  query: string
  embedding_model: string
  latency_ms: number
  items: RetrievedChunk[]
}
